### PR TITLE
implement change password in user profile

### PR DESCRIPTION
**Pull request by petrjasek**
in case the auth provider supports it.
also move the firebase auth integration to core.

CPCN-206

## Checklist

- [ ] This pull request is not adding new forms that use redux
- [ ] This pull request is adding missing TypeScript types to modified code segments where it's easy to do so with confidence
- [ ] This pull request is replacing `lodash.get` with optional chaining for modified code segments

### CODE REVIEW

File: assets/user-profile/components/profile/UserProfile.tsx [hunk 1]
```diff
@@ -202,17 +217,21 @@ class UserProfile extends React.Component<IProps, any> {
...
 const mapDispatchToProps = (dispatch: any) => ({
     saveUser: () => dispatch(saveUser()),
     fetchUser: (userId: IUser['_id']) => dispatch(fetchUser(userId)),
-    onChange: (event: React.ChangeEvent) => dispatch(editUser(event)),
+    onChange: (updates: IUserProfileUpdates) => dispatch(editUser(updates)),
     setError: (errors: {[key: string]: string}) => dispatch(setError(errors)),
     openEditTopicNotificationsModal: () => dispatch(openEditTopicNotificationsModal()),
 });
 
+type IStateProps = typeof mapStateToProps;
```
Comment (MarkLark86) [inline assets/user-profile/components/profile/UserProfile.tsx:234]: These 2 types aren't used anywhere. Also, the version of Redux we're using doesn't play nicely with typescript (I have updated it in my multiday events PR to the newest version that does)

File: newsroom/auth/providers.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import abc
+
+# from type
```
Comment (MarkLark86) [inline newsroom/auth/providers.py:3]: Commented line not needed

File: newsroom/auth/views.py [hunk 3]
```diff
@@ -283,3 +308,60 @@ def impersonate_stop():
...
+        if auth_provider.type == AuthProviderType.FIREBASE:
+            if form.data.get("firebase_status"):
+                firebase_status = form.data["firebase_status"]
+                if firebase_status == "OK":
+                    flask.flash(gettext("Your password has been changed."), "success")
+                elif firebase_status == "auth/wrong-password":
+                    flask.flash(gettext("Wrong current password."), "error")
+                else:
+                    logger.warning("Unhandled firebase error %s", firebase_status)
+                    flask.flash(gettext("Could not change password. Please try again later."), "warning")
```
Comment (MarkLark86) [inline newsroom/auth/views.py:332]: I think `Please try again later` might not be good wording here. The user would be thinking if it doesn't work now, how could it work later.
What if we instead suggest to them to contact their administrator?

File: assets/auth/firebase/reset_password.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        const data = new FormData(form);
+        const email = data.get('email') as string;
+
+        params.append('email', email);
+        url.search = params.toString();
+
+        sendButton.disabled = true;
+        sendPasswordResetEmail(auth, email, {url: url.toString()})
+            .then(() => {
+                // set `email_sent` to true, so server knows password reset was handled by firebase
```
Comment (MarkLark86) [inline assets/auth/firebase/reset_password.ts:28]: I see `email_sent` is no longer used, as we're checking the auth provider on the server side. Therefor this comment is no longer required

**Review by MarkLark86 — APPROVED** [review]
Looks fine to me, only a couple small things

File: assets/auth/firebase/change_password.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        const data = new FormData(form);
+        const oldPassword = data.get('old_password') as string;
+        const newPassword = data.get('new_password') as string;
+        const newPassword2 = data.get('new_password2') as string;
+
+        if (newPassword !== newPassword2) {
+            form.submit();
+        }
+
+        signInWithEmailAndPassword(auth, firebaseUserEmail, oldPassword).then((userCredential) => {
```
[the remark this answers is not in the document]
Comment (petrjasek) [reply assets/auth/firebase/change_password.ts:22]: no, but will add there return after submit to make it more explicit

**Review by MarkLark86 — APPROVED** [review]

**Pull request merged**